core/state/statedb.go

```diff
@@ -1297,20 +1360,47 @@ func (s *StateDB) Commit(failPostCommitFunc func(), postCommitFuncs ...func() er
 	var diffLayer *types.DiffLayer
 	var verified chan struct{}
 	var snapUpdated chan struct{}
+	var snapCreated chan common.Hash
+	var accountRefreshed chan struct{}
 	if s.snap != nil {
 		diffLayer = &types.DiffLayer{}
 	}
 	if s.pipeCommit {
 		// async commit the MPT
 		verified = make(chan struct{})
 		snapUpdated = make(chan struct{})
+		snapCreated = make(chan common.Hash)
+		accountRefreshed = make(chan struct{})
 	}
 
 	commmitTrie := func() error {
 		commitErr := func() error {
+			accountData := make(map[common.Hash][]byte)
+			if s.pipeCommit && s.snap != nil {
+				// Due to state verification pipeline, the accounts roots are not updated, leading to the data in the difflayer is not correct
+				// Fix the wrong data here
+				accountData = s.accountDataForDiffLayer()
```

Comment:
`accountData = s.accountDataForDiffLayer()` is needed no matter `s.pipeCommit` `s.snap` is false or nil.